lib/srv/reexec.go

```diff
@@ -1202,48 +1207,76 @@ func copyCommand(ctx *ServerContext, cmdmsg *ExecCommand) {
 	}
 }
 
-// HasAccessibleHomeDir checks if the given User has access to an existing home directory.
-func HasAccessibleHomeDir(localUser *user.User) error {
-	fi, err := os.Stat(localUser.HomeDir)
+func coerceHomeDirError(usr *user.User, err error) error {
+	if os.IsNotExist(err) {
+		return trace.NotFound("home directory %q not found for user %q", usr.HomeDir, usr.Name)
+	}
+
+	if os.IsPermission(err) {
+		return trace.AccessDenied("%q does not have permission to access %q", usr.Name, usr.HomeDir)
+	}
+
+	return err
+}
+
+// hasAccessibleHomeDir checks if the current user has access to an existing home directory.
+func hasAccessibleHomeDir() error {
+	// this should usually be fetching a cached value
+	currentUser, err := user.Current()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	fi, err := os.Stat(currentUser.HomeDir)
+	if err != nil {
+		return trace.Wrap(coerceHomeDirError(currentUser, err))
+	}
+
 	if !fi.IsDir() {
-		return trace.NotFound("user home is not a directory")
+		return trace.BadParameter("%q is not a directory", currentUser.HomeDir)
 	}
 
-	stat, ok := fi.Sys().(*syscall.Stat_t)
-	if !ok {
-		return trace.AccessDenied("could not retrieve owner of home directory")
+	executable, err := os.Executable()
+	if err != nil {
+		return trace.Wrap(err)
 	}
 
-	uid := strconv.Itoa(int(stat.Uid))
-	gid := strconv.Itoa(int(stat.Gid))
+	// attemping to cd into the target directory is the easiest, cross-platform way to test
+	// whether or not the current user has access
+	cwd := filepath.Dir(executable)
+	if err := os.Chdir(currentUser.HomeDir); err != nil {
+		return trace.Wrap(coerceHomeDirError(currentUser, err))
+	}
 
-	if uid == localUser.Uid && gid == localUser.Gid {
-		return nil
+	if err := os.Chdir(cwd); err != nil {
+		return trace.Errorf("unable to return to original working directory")
 	}
 
-	return trace.AccessDenied("user does not own configured home directory")
+	return nil
 }
 
-// CheckHomeDir checks if the user's home dir exists and is accessible. This also handles cases where
-// the home directory isn't visible to the root user, which HasAccessibleHomeDir doesn't account for.
-func CheckHomeDir(localUser *user.User) (bool, error) {
-	err := HasAccessibleHomeDir(localUser)
-	if err == nil {
-		return true, nil
+func CheckHomeDirAccess(localUser *user.User) (bool, error) {
+	currentUser, err := user.Current()
+	if err != nil {
+		return false, trace.Wrap(err)
 	}
 
-	if trace.IsAccessDenied(err) {
-		return false, nil
+	// don't spawn a subcommand if already running as the user in question
+	if currentUser.Uid == localUser.Uid {
```

Comment:
It seems like there was an assumption baked in that `CheckHomeDir` would always be called by the root/primary user, but I didn't observe that during testing. In fact I never actually saw the subcommand getting spawned while testing ssh connections through the web UI 🤔 This check prevents dropping into another sub process if we're already running as the target user